django/core/management/commands/squashmigrations.py

```diff
@@ -157,10 +151,7 @@ def handle(self, **options):
         # need to feed their replaces into ours
         replaces = []
         for migration in migrations_to_squash:
-            if migration.replaces:
-                replaces.extend(migration.replaces)
-            else:
-                replaces.append((migration.app_label, migration.name))
```

Comment:
I checked your regression test; it fails on `main`, as it should. However, when applying _only_ your changes to the `squashmigrations.py` command (here) and _not_ your changes to the migration loader, your test case passes.

Reply:
Alright, I looked into this a bit today. 

So my main dilemma is that, at least running locally, the order of migration loading (for disk migrations) is non-deterministic in the old version of the loader. In attempting to make a failing test for this, I found that the existing test _does_ fail sometimes with the old code without the migration loader. But only sometimes! 

It's dependent on the ordering of the iteration of the migrations when the disk migrations are loaded in `load_disk`. In particular the names are loaded into a set (that is where the non-determinism comes I think, since dictionary iteration should in theory be stable over runs). 

So I think it's not really possible to make a test case that would fail consistently with the old code, while being a correct migration graph in the new model.

Would it be good to say that the added `test_squashmigrations_squashes_already_squashed` test (which fails non-deterministically with the old logic) covers the main logic and the "happy path", but that I still should add one test for the cycle tracking in the replacement logic?

Reply:
I'm just getting up to speed again, but I do see that reverting the loader changes now leads to a consistent failure (good):

```shell
django.db.migrations.exceptions.NodeNotFoundError: Unable to find replacement node ('migrations', '3_squashed_5'). It was either never added to the migration graph, or has been removed.
```